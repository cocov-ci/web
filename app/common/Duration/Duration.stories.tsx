import type { Meta, StoryObj } from '@storybook/react'

import Duration from './index'

const meta: Meta<typeof Duration> = {
  title: 'Common/Duration',
  component: Duration,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof Duration>

export const Default: Story = {
  args: {
    fromDate: '01 Jan 1970 00:00:00 GMT',
    toDate: '01 Jan 1970 00:00:03 GMT',
  },
}

export const OneMinute: Story = {
  args: {
    fromDate: '01 Jan 1970 00:00:00 GMT',
    toDate: '01 Jan 1970 00:01:00 GMT',
  },
}

export const OneHour: Story = {
  args: {
    fromDate: '01 Jan 1970 00:00:00 GMT',
    toDate: '01 Jan 1970 01:00:00 GMT',
  },
}

export const OneMinuteTwoSeconds: Story = {
  args: {
    fromDate: '01 Jan 1970 00:00:00 GMT',
    toDate: '01 Jan 1970 00:01:02 GMT',
  },
}

export const OneHourOneSecond: Story = {
  args: {
    fromDate: '01 Jan 1970 00:00:00 GMT',
    toDate: '01 Jan 1970 01:00:01 GMT',
  },
}

export const TwoHoursOneMinuteThreeSeconds: Story = {
  args: {
    fromDate: '01 Jan 1970 00:00:00 GMT',
    toDate: '01 Jan 1970 02:01:03 GMT',
  },
}

export const Zero: Story = {
  args: {
    fromDate: '01 Jan 1970 00:00:00 GMT',
    toDate: '01 Jan 1970 00:00:00 GMT',
  },
}
