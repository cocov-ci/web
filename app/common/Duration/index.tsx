import classNames from 'classnames'
import React from 'react'

interface DurationProps {
  fromDate: string
  toDate?: string
  className?: string
}

const pluralize = (quantity: number, singular: string) =>
  `${quantity} ${singular}${quantity > 1 ? 's' : ''}`

const Duration = ({ className, fromDate, toDate }: DurationProps) => {
  const parsedFrom = Date.parse(fromDate)
  let parsedTo: number

  if (toDate) {
    parsedTo = Date.parse(toDate)
  } else {
    parsedTo = Date.now()
  }

  const duration = Math.abs(parsedTo - parsedFrom) / 1000.0
  const hours = ~~(duration / 3600)
  const minutes = ~~((duration % 3600) / 60)
  const seconds = ~~duration % 60
  const components = []

  if (hours > 0) {
    components.push(pluralize(hours, 'hour'))
  }

  if (minutes > 0) {
    components.push(pluralize(minutes, 'minute'))
  }

  if (seconds > 0) {
    components.push(pluralize(seconds, 'second'))
  }

  if (hours + minutes + seconds === 0) {
    components.push('less than a second')
  }

  let result

  if (components.length > 1) {
    const last = components.pop()
    result = `${components.join(', ')} and ${last}`
  } else {
    result = `${components.pop()}`
  }

  return <span className={classNames(className)}>{result}</span>
}

export default Duration
