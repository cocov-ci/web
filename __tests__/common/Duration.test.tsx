import { render, screen } from '@testing-library/react'

import Duration from 'app/common/Duration'

import '@testing-library/jest-dom'

const args = {
  fromDate: '01 Jan 1970 00:00:00 GMT',
  toDate: '01 Jan 1970 00:00:03 GMT',
}

describe('common/Duration', () => {
  it('renders Duration', () => {
    render(<Duration {...args} />)

    expect(screen.getByText('3 seconds')).toBeVisible()
  })

  it('renders Duration passing only FromDate', () => {
    render(
      <Duration
        fromDate={(() => {
          const d = Date.now() - 3 * 60 * 1000

          return new Date(d).toISOString()
        })()}
      />,
    )

    expect(screen.getByText('3 minutes')).toBeVisible()
  })

  it('renders Duration higher than 1 hour', () => {
    const args = {
      fromDate: '01 Jan 1970 00:00:00 GMT',
      toDate: '01 Jan 1970 01:10:00 GMT',
    }

    render(<Duration {...args} />)

    expect(screen.getByText('1 hour and 10 minutes')).toBeVisible()
  })

  it('renders Duration less than a second', () => {
    const args = {
      fromDate: '01 Jan 1970 00:00:00 GMT',
      toDate: '01 Jan 1970 00:00:00 GMT',
    }

    render(<Duration {...args} />)

    expect(screen.getByText('less than a second')).toBeVisible()
  })

  it('renders Duration snapshots', () => {
    const { container } = render(<Duration {...args} />)

    expect(container).toMatchSnapshot()
  })
})
