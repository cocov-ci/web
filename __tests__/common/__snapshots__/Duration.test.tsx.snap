// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Duration renders Duration snapshots 1`] = `
<div>
  <span
    class=""
  >
    3 seconds
  </span>
</div>
`;
